src/Common/RemoteProxyConfigurationResolver.cpp

```diff
@@ -16,20 +16,63 @@ namespace ErrorCodes
     extern const int BAD_ARGUMENTS;
 }
 
+std::string RemoteProxyHostFetcherImpl::fetch(const Poco::URI & endpoint, const ConnectionTimeouts & timeouts) const
+{
+    /// It should be just empty GET request.
+    Poco::Net::HTTPRequest request(Poco::Net::HTTPRequest::HTTP_GET, endpoint.getPath(), Poco::Net::HTTPRequest::HTTP_1_1);
+
+    const auto & host = endpoint.getHost();
+    auto resolved_hosts = DNSResolver::instance().resolveHostAll(host);
+
+    HTTPSessionPtr session;
+
+    for (size_t i = 0; i < resolved_hosts.size(); ++i)
+    {
+        auto resolved_endpoint = endpoint;
+        resolved_endpoint.setHost(resolved_hosts[i].toString());
+        session = makeHTTPSession(HTTPConnectionGroupType::HTTP, resolved_endpoint, timeouts);
+
+        try
+        {
+            session->sendRequest(request);
+            break;
+        }
+        catch (...)
+        {
+            if (i + 1 == resolved_hosts.size())
+                throw;
+        }
+    }
+
+    Poco::Net::HTTPResponse response;
+    auto & response_body_stream = session->receiveResponse(response);
+
+    if (response.getStatus() != Poco::Net::HTTPResponse::HTTP_OK)
+        throw Exception(ErrorCodes::BAD_ARGUMENTS, "Proxy resolver returned not OK status: {}", response.getReason());
+
+    String proxy_host;
+    /// Read proxy host as string from response body.
+    Poco::StreamCopier::copyToString(response_body_stream, proxy_host);
```

Comment:
that read also could throw exception.

Reply:
Do we need to catch it tho? I think it's ok to throw the exception in case fetching the proxy host fails. I wouldn't want to have a proxy server configured that is bypassed when it fails. What do you think?

Reply:
We need to answer to the question: "who is responsible for making retries here?".
Do AWS SDK retry proxy resolve call? If so, than it is enough and we should do any retry here, just pass the error up.
If anyone is retry that proxy resolve calls, that we have to do it here.

There is a possibility to get just network error in each call which has been mentioned above. That errors are not fatal, they have to be retried.

Reply:
Ok, AFAIK Poco does not implement retry logic by default. I can implement a simple one with a fixed number of retries.

Reply:
Try to use ReadWriteBufferFromHTTP. It does retries.

Reply:
Done

Reply:
Is there a way I can check for the http status code? I can't find an API that allows me to do that

Reply:
ReadWriteBufferFromHTTP checks it. If it bad, the request is either retried or an excpection is thrown.

Reply:
Ok, there has been a small confusion. When you asked me if AWS takes cares of retries for proxy calls, I thought you were asking about the client we use to make the proxy calls, that's why I said no retries were in place.

AWS indeed retries this. If I add retry logic to proxy retrieval, we get retries of retries - we should avoid it. Even if AWS did not retry, I don't think we should implement retries here. I think this should be up to the caller, otherwise it might be "uncontrollable".

Therefore, I suggest we just keep it simple: either `makeHTTPSession` or `ReadWriteBufferFromHTTP` with max_http_Retry=1

Reply:
I think it is a little bit more tricky.

Here we are mutting all the errors from http request to the resolver. If request failed we return empty config.
https://github.com/ClickHouse/ClickHouse/blob/f82a816fdd9de00788efa15dac5b1881a8141a11/src/Common/RemoteProxyConfigurationResolver.cpp#L113-L117

I do not understand that. Why we return empty config? Why not cached config from previous resolve request? Or why not rethrow all exception up to trigger retry request?

How is it suppose to work if we have just unstable network here?

Reply:
I keep taking this code for granted, but I agree we should let the exception propagate to hit the retry logic